LiveSPICEVst/LiveSPICEPlugin.cs

```diff
@@ -0,0 +1,304 @@
+
+            // We only have one program
+
+            var program = new Program();
+            program.Name = null; // "Program 1";
+            program.Data = null;
+
+            VstProgramParameters programParameters = new VstProgramParameters
+            {
+                SchematicPath = SimulationProcessor.SchematicPath,
+                OverSample = SimulationProcessor.Oversample,
+                Iterations = SimulationProcessor.Iterations
+            };
+
+            foreach (ComponentWrapper wrapper in SimulationProcessor.InteractiveComponents)
+            {
+                if (wrapper is PotWrapper)
+                {
+                    programParameters.ControlParameters.Add(new VSTProgramControlParameter { Name = wrapper.Name, Value = (wrapper as PotWrapper).PotValue });
+                }
+                else if (wrapper is ButtonWrapper)
+                {
+                    programParameters.ControlParameters.Add(new VSTProgramControlParameter { Name = wrapper.Name, Value = (wrapper as ButtonWrapper).Engaged ? 1 : 0 });
+                }
+            }
+
+            XmlSerializer serializer = new XmlSerializer(typeof(VstProgramParameters));
+
+            using (MemoryStream memoryStream = new MemoryStream())
+            {
+                serializer.Serialize(memoryStream, programParameters);
+
+                program.Data = memoryStream.ToArray();
+            }
+
+            return program;
+        }
+
+        /// <summary>
+        /// Set plugin state from program data
+        /// </summary>
+        /// <param name="program">The program data</param>
+        /// <param name="index">The index of the program</param>
+        public void SetProgramData(Program program, int index)
+        {
+            Logging.Log("Load program: " + index);
+
+            XmlSerializer serializer = new XmlSerializer(typeof(VstProgramParameters));
+
+            try
+            {
+                using (MemoryStream memoryStream = new MemoryStream(program.Data))
+                {
+                    VstProgramParameters programParameters = serializer.Deserialize(memoryStream) as VstProgramParameters;
+
+                    if (string.IsNullOrEmpty(programParameters.SchematicPath))
+                    {
+                        haveSimulationError = false;
+
+                        SimulationProcessor.ClearSchematic();
+                    }
+                    else
+                    {
+                        LoadSchematic(programParameters.SchematicPath);
+                    }
+
+                    SimulationProcessor.Oversample = programParameters.OverSample;
+                    SimulationProcessor.Iterations = programParameters.Iterations;
+
+                    foreach (VSTProgramControlParameter controlParameter in programParameters.ControlParameters)
+                    {
+                        foreach (ComponentWrapper wrapper in SimulationProcessor.InteractiveComponents)
```

Comment:
I've never really adopted LINQ in my coding, although I think it is cool. Not 100% sure it makes this clearer, but I went ahead a switched it.

I did the null check, but I don't think it is worth generating an error. It would happen if you removed or renamed a pot or button and had it referenced in a previously saved preset.

Reply:
That makes sense, I agree this doesn't make sense to generate an error for.